[1.21.4] Tear out `ExistingFileHelper`

ApexModder opened the pull request:
Follow up PR from #1725 which aims to deprecate `ExistingFileHelper` (EFH) further with no planned replacement.

This is done by marking EFH as `@Nullable` in as many places as possible and providing overloaded constructors to allow constructing providers without needing to pass a EFH instance.

All NeoForge and test providers have been updated to use the new overloaded constructors, this is to validate that everything still works when no EFH instance is provided.

ApexModder committed:
Deprecate `ExistingFileHelper` and mark nullable everywhere

Remove usages of `ExistingFileHelper` is as many neoforge/test data providers as possible

ApexModder committed:
Tear out `ExistingFileHelper`

ApexModder commented:
PR has been updated to fully tear out `ExistingFileHelper` rather then deprecating it for removal

KnightMiner commented:
I'm just going to go on record saying I extensively use the existing file helper to run datagen based on existing files. Be it taking in 1 or more existing images and generating a new image based on them, or taking in existing structure NBT and modifying it for output, or even in some cases taking a JSON file and manipulating it for a generated file.

Removing this API would break every single one of those, you need to provide a replacement unless you are just planning to piss off modders again by removing APIs we are actively using. I am not the only one making use of existing files for datagen, it's useful for far more than checking if a file is present.

We don't need anything fancy, just needs to be able to fetch a resource from either our own resources, vanilla resources, or a mod dependency resources

Minecraftschurli reviewed (dismissed):
Query the community for if this is still being used and requires a replacement
(like @KnightMiner is saying he makes extensive use of it)

ApexModder committed:
Remove old `@Deprecated` marker

ApexModder committed:
Add super simple `ResourceManager` implementation

Add back `assetIndex`/`assetsDir` to `clientData` run config args
Add back `DataGeneratorTest` sound/particle tests making use of new resource manager

ApexModder commented:
Added a very simple `ResourceManager` like class
which allows looking up vanilla/modded resources at runtime during datagen
it does not track generated resources however, only existing files on disk afaik
and does not require modders to specify `--existing` or `--existing-mod`
(simply adds vanilla and all loaded mods as valid pack sources)

structure is like so, all of these methods function exactly like any regular `ResourceManager`
only change is they require a `PackType` param to know if its a client or server resource
```java
public interface DataResourceManager {
    public Optional<Resource> getResource(PackType packType, ResourceLocation path);
    public boolean exists(PackType packType, ResourceLocation path);
    public Resource getResourceOrThrow(PackType packType, ResourceLocation path) throws FileNotFoundException;
    public InputStream open(PackType packType, ResourceLocation path) throws IOException;
    public BufferedReader openAsReader(PackType packType, ResourceLocation path) throws IOException;
    public List<Resource> getResourceStack(PackType packType, ResourceLocation path);
    public Map<ResourceLocation, Resource> listResources(PackType packType, String path, Predicate<ResourceLocation> filter);
    public Map<ResourceLocation, List<Resource>> listResourceStacks(PackType packType, String path, Predicate<ResourceLocation> filter);
}
```

this is stored on the `GatherDataEvent` much like EFH was and can simply be passed around as needed
hopefully this serves knights (and other modders) usages

XFactHD commented on src/main/java/net/neoforged/neoforge/common/data/AdvancementProvider.java:
In absence of the `ExistingFileHelper`, this class provides no benefit over the vanilla `AdvancementProvider` and the associated `AdvancementSubProvider` and should thus be removed.

XFactHD commented on src/main/java/net/neoforged/neoforge/common/data/SoundDefinitionsProvider.java:
```diff
@@ -200,7 +196,9 @@ private boolean validate(final String name, final SoundDefinition def) {
     private boolean validate(final String name, final SoundDefinition.Sound sound) {
```
This method and the `validateEvent()` method can be condensed into the `validate()` method above by filtering the stream for `SoundType.EVENT` first and folding `validateEvent()` into the `allMatch()`.

XFactHD commented on src/main/java/net/neoforged/neoforge/data/event/GatherDataEvent.java:
```diff
@@ -28,25 +28,27 @@
         return this.modContainer;
     }
 
+    public DataResourceManager getResourceManager() {
```
This method should take a `PackType` and return the respective `ResourceManager` directly instead of returning a custom type that basically just duplicates a bunch of methods for very little benefit.

XFactHD commented on src/main/java/net/neoforged/neoforge/common/data/DataResourceManager.java:
```diff
@@ -0,0 +1,108 @@
+
+        server.push(ServerPacksSource.createVanillaPackSource());
+
+        ModList.get().forEachModInOrder(mod -> {
+            var packInfo = new PackLocationInfo("mod/" + mod.getModId(), Component.empty(), PackSource.BUILT_IN, Optional.empty());
+            var modPack = ResourcePackLoader.createPackForMod(mod.getModInfo().getOwningFile()).openPrimary(packInfo);
+
+            client.push(modPack);
+            server.push(modPack);
+        });
```
This should filter out the Minecraft "mod" like `ResourcePackLoader.findResourcePacks()` does and likely also needs to filter out the mod for which datagen is being run to avoid including data/assets which will be overwritten by this run (may require adding back the `--existing` parameter).

XFactHD commented on src/main/java/net/neoforged/neoforge/common/data/DataResourceManager.java:
```diff
@@ -0,0 +1,108 @@
+ * @see ResourceManager
+ */
+public final class DataResourceManager {
+    private final FallbackResourceManager client = new FallbackResourceManager(PackType.CLIENT_RESOURCES, "minecraft");
+    private final FallbackResourceManager server = new FallbackResourceManager(PackType.SERVER_DATA, "minecraft");
```
These need to be `MultiPackResourceManager`s as `FallbackResourceManager`s only support a single namespace.

Technici4n commented:
Here is an example of accessing resources for usage in datagen: https://github.com/AztechMC/Modern-Industrialization/blob/5d9c1b69f3d6d683554f2003f6c688b89909e676/src/client/java/aztech/modern_industrialization/datagen/texture/TexturesProvider.java#L73. I think it's really simple, and doesn't require any specific NeoForge extension. @KnightMiner

ApexModder committed:
Split `DataResourceManager` into 2 instances per `PackType`

ApexModder committed:
Drop `DataResourceManager` class type

Construct data resource managers directly as `MultiPackResourceManager`

ApexModder committed:
Drop our `AdvancementProvider` serves no use with out `ExistingFileHelper`

Use vanillas `AdvancementProvider` instead now

ApexModder committed:
Condense down `SoundDefinitionsProvider` validation method

ApexModder committed:
Revive `--existing`

Minecraftschurli reviewed (approved):
:shipit:
